### ISSUE TITLE
Data

### ISSUE THREAD

**Issue by ibmua** [ltrottier/pelu.resnet.torch#1]
Great study! I'm going to try to use this activation type. =) 

Just a few questions:
1)  "Global pixel-wise mean subtraction" = global mean subtraction, or pixel-wise mean subtraction as in measuring distinct mean of each pixel along training data (which should be awful from theoretical conv net perspective, but is one of the standard things in Caffe, I belive)? Edit: from your code, I understand that it's the first one, which is good =)

2) Have you tried applying this thing per channel (feature map)? I've seen that you wrote you were "trying to prevent overfitting", but it's not like fitting the data is a bad thing, it's actually very good and doing it via parameters of activation functions may be even "safer" in terms of learning the right thing than doing it via getting a wider/longer network and is likely more efficient in terms of flops than those measures.

**Comment by ibmua** [comment]
Looking again at this after trying PELUs, yeah, it's hugely overfitting. But it would still be interesting to test per-channel, though, anyway.

I think, finding a really good regularization for this thing is of utmost importance. This thing is really cool, but it needs to be tamed well to reap it's benefits. I think, per-channel parametrization would do great things once a proper regularization method is discovered.

**Comment by ltrottier** [comment]
Hi,

Thanks for the comment.

To answer your second question, I did not apply it channel-wise. Maybe I'll do it if I have time. And concerning over-fitting, I had to increase the weight decay from 0.0001 to 0.001 to get the results. Did you try other weight decays ?

**Comment by ibmua** [comment]
Yeah, but the architecture on which I've tried that may have been prone to overfitting. I'll try it with other ones too. I'll report if I'll get any cool findings, or results. (>‿◠)✌

**Comment by ibmua** [comment]
BTW, I didn't get deep into the algo, but is there maybe a likewise regularization for the PELU fitting kind of like weight decay? Pelu_weight_update = Pelu_weight_update - decay\* (Pelu_weight_update^2) \* sign(Pelu_weight_update) , something like that? Maybe that would add some regularization from another side?

My guess is that PELU provides a good way for a smaller # of weights to fit the data. But without proper adjustments, the # of weights is actually smaller than it should be. So, maybe, making PELU fitting slower, more amortized, maybe, may probably give the weights more time to adjust correctly.

**Comment by ibmua** [comment]
My intuition is that there's some level of elastic plasticities that is highly efficiently beneficial, but it is right a big question how to make it so that learning those plasticities doesn't impede learning conceptual relationships. My intuition is that it's likely that exact nature of activation should probably be learned further and further during training and that it has to learn them very slowly compared to how conceptual relationships that weights are would be adapt. So I'd say that at the beginning activation functions should be something like a plain ReLU and adaptively become more like PELU and APL very far during learning with the sort of LR kind of like we have in weights, except probably more of the reverse order. Start out as a very very slow almost non-plastic learner and get to an almost liquid state with time. 

Something gives me a pretty high certainty that's the missing ingredient in all of current activation functions, the key to produce the ultimate activation function. Wonder if anyone figured that out already, because so far I usually found that all my NN ideas were most of the time published as a paper even before I started exploring them, though very recently and so not very searchable.

**Comment by ibmua** [comment]
Notice that this would also speed up learning by a lot. As in training time speed up due to use of ReLU, or some LReLU at the beginning.

**Comment by ltrottier** [comment]
Let's go back for a moment to the basics of what a neural network does. To put it simply, it starts by computing a = Wx, where a is a vector containing, in its elements, dot products between observation x and features w_i. So, a_i = < w_i, x>. A dot product can be seen as a non-normalized cosine similarity. If a_i  is close to zero, the feature w_i and the observation x are near orthogonal, and when a_i has a high magnitude (either positive or negative), x and w_i are near collinear. So all a network does in the first part is to look at linear correlations. Then comes the activation function. You can see the activation function as a rule for processing these correlations. In the case of ReLU for instance, it keeps positive correlations and removes negative ones. For ELU, it also keeps positive correlations but saturates negatives correlations at -1. In the case of adaptive activation function, the network learns what to do with these correlations.

Of course since neural networks are universal approximators, all you need is a Sigmoid. But in reality, we have to look at other considerations so that we can learn something useful. For instance, because we use gradient descent to learn the weights, we have to be careful about vanishing gradients. Or, we have to initialize the weights in a certain way because we use a certain activation function so that we don't have diminishing feature reuse. 

So in my opinion, there won't be an ultimate activation function. It will only be a function with certain properties that will fit with the rest. Perhaps we will discover a new way to learn the weights that won't involve gradients, and in that particular case the sigmoid will be more suitable, even though it makes the gradients vanish (but we won't care at that time). But that's just my honest opinion.

**Comment by ibmua** [comment]
I wonder, isn't PELU's curve nearly identical to softplus? Just the curve, I understand that unlike softplus, it goes below 0.

**Comment by ibmua** [comment]
Have you tried numeric datasets? My bet is this thing will most likely perform with a much higher margin than what you'll be able to get out of it with images. You see in image recognition it's not much about regression as far as my tests go. If you go the regression way you'll be able to match the training data, but without much meaningful pattern extraction, it seems. The problem is that training and testing data are way too different in images, so you have to seriously constrain it to have a pure reasoning. And plain ReLUs/Maxout and, possibly, also plain linear activation, seem to be very good at such constraining. But with tasks that are inherently about regression and smooth functions PELU and your general way of reasoning about NNs should perform great, as far as my guesses and previous experiments with ELU on regression go. 

Though in some places in those networks, most likely closer to their beginnings, it may be beneficial to still use some Linear/ReLUs/PReLUs/Maxout aswell to constrain their reasoning to more clearly logical and so less overfitting functions.

I urge you to try numeric datasets. I feel strongly that PELU's going to show much more benefit there.

**Comment by ibmua** [comment]
My guess regarding weight decay increase is that when your gradient has a `x²`-like curve, weight decay worked in a `error => ⁿ√error` kind of way (I guess, similar to root-mean-square error logic?) pushing the exaggerations caused by `x²`-like curve down. Something like that. Was that the idea?

BTW, I was wondering about this thing. People use weight decay calculated as `e = e - k*e²` for faster calculations, though the much more correct way would be using `error => ⁿ√error`, right? Because with `e = e - k*e²` some values that were originally larger can easily become smaller than the values that were originally smaller and it overall looks extremely harsh, especially for larger `k` values.

**Comment by ibmua** [comment]
My guess is PELU should shine most when used to regress huge extensive numeric datasets with not-so-large networks.

**Comment by ibmua** [comment]
Okay, so I've found that wrong weight decay formula when searching the web. But the actual WD used in optim Torch module is very a different one.

**Comment by ibmua** [comment]
Maybe, it's also gonna work nicely in segmentation.
